web/src/main/java/com/microsoft/applicationinsights/web/internal/correlation/TraceContextCorrelation.java

```diff
@@ -0,0 +1,448 @@
+        .getInstrumentationKey());
+
+        String target = resolve(headerAppID, currAppId);
+        if (target == null) {
+            InternalLogger.INSTANCE.warn("Target value is null and hence returning empty string");
+            return ""; // we want an empty string instead of null so it plays nicer with bytecode injection
+        }
+        return target;
+    }
+
+    /**
+     * Extracts the appId/roleName out of Tracestate and compares it with the current appId. It then
+     * generates the appropriate source or target.
+     */
+    private static String generateSourceTargetCorrelation(String instrumentationKey, String appId) {
+
+        assert instrumentationKey != null;
+        assert appId != null;
+
+        String myAppId = InstrumentationKeyResolver.INSTANCE.resolveInstrumentationKey(instrumentationKey);
+
+        return resolve(appId, myAppId);
+    }
+
+    /**
+     * Resolves appId based on appId passed in header and current appId
+     * @param headerAppId
+     * @param currentAppId
+     * @return
+     */
+    private static String resolve(String headerAppId, String currentAppId) {
+
+        //it's possible the appId returned is null (e.g. async task is still pending or has failed). In this case, just
+        //return and let the next request resolve the ikey.
+        if (currentAppId == null) {
+            InternalLogger.INSTANCE.trace("Could not generate source/target correlation as the appId could not be resolved (e.g. task may be pending or failed)");
+            return null;
+        }
+
+        // if the current appId and the incoming appId are send null
+        String result = null;
+        if (headerAppId != null && !headerAppId.equals(currentAppId)) {
+            result = headerAppId;
+        }
+
+        return result;
+    }
+
+    /**
+     * Helper method to retrieve Tracestate from ThreadLocal
+     * @return
+     */
+    public static String retriveTracestate() {
+        //check if context is null - no correlation will happen
+        if (ThreadContext.getRequestTelemetryContext() == null || ThreadContext.getRequestTelemetryContext().
+                getTracestate() == null) {
+            InternalLogger.INSTANCE.warn("No correlation wil happen, Thread context is null");
+            return null;
+        }
+
+        Tracestate tracestate = ThreadContext.getRequestTelemetryContext().getTracestate();
+        return tracestate.toString();
+    }
+
+    /**
+     * Generates child TraceParent by retrieving values from ThreadLocal.
+     * @return Outbound Traceparent
+     */
+    public static String generateChildDependencyTraceparent() {
+        try {
+
+            RequestTelemetryContext context = ThreadContext.getRequestTelemetryContext();
+
+            //check if context is null - no correlation will happen
+            if (context == null) {
+                InternalLogger.INSTANCE.warn("No Correlation will happen, Thread context is null while generating child dependency");
```

Comment:
why? if there is no parent context, we should still generate a new traceparent and add to the headers. This would allow us to trace dependency to the request on the next service and build at least appmap.

Reply:
The previous correlation worked this way atleast, assuming that there is one incoming request. This is a valid scenario though. A worker role can initiate an outbound call. I will update this one.

Reply:
Done